inst/examples/shiny_interactions.R

```diff
@@ -0,0 +1,247 @@
+    if(is.null(input$addfeature)==FALSE && input$addfeature != "") {
+      geojson <- geojson$features[[seq_along(geojson$features)[sapply(geojson$features,
+              FUN = function(x) x[["id"]] == input$addfeature)]]]
+      leafletProxy("mymap") %>%
+      addFeatureGeoJSON(geojson, layerId ='geojsonlayer') # can use a list (slow)
+      if (length(addedData$df$ids) > 1) {
+        removedData$df <- removedData$df[-c(which(removedData$df$ids==input$addfeature)),]
+      }
+      addedData$df <- rbind(addedData$df,allCountries[which(allCountries$ids==input$addfeature),])
+      updateSelectizeInput(session, 'setstyle', choices = addedData$df$ids, server = TRUE, selected=NULL)
+      updateSelectizeInput(session, 'removefeature', choices = addedData$df$ids, server = TRUE, selected=NULL)
+      updateSelectizeInput(session, 'addfeature', choices = removedData$df$ids, server = TRUE, selected=NULL)
+    }
+  })
+  observeEvent(input$clearGeojson, {
+    leafletProxy("mymap") %>% removeGeoJSON(layerId ='geojsonlayer')
+  })
+  observeEvent(input$addPopup, {
+    content <- paste(sep = "<br/>",
+                     "<b><a href='http://www.samurainoodle.com'>Samurai Noodle</a></b>",
+                     "606 5th Ave. S",
+                     "Seattle, WA 98138"
+    )
+    leafletProxy("mymap") %>%  addPopups(-122.327298, 47.597131, content,
+                         options = popupOptions(closeButton = TRUE))
+  })
+  observeEvent(input$clearPopup, {
+    leafletProxy("mymap") %>% clearPopups()
+  })
+  observeEvent(input$addbasemap, {
+    leafletProxy("mymap") %>% addProviderTiles("Acetate.terrain",options = providerTileOptions(noWrap = TRUE,zIndex=0),layerId="basemap")
+  })
+  observeEvent(input$clearbasemap, {
+    leafletProxy("mymap") %>% removeTiles("basemap") %>% removeTiles("wms_protectedArea")
+  })
+  observeEvent(input$clear, {
+    leafletProxy("mymap") %>% clearTiles()
+  })
+  observeEvent(input$mymap_geojson_mouseover, {
+        leafletProxy("mymap") %>%
+          styleFeatureGeoJSON(layerId ='geojsonlayer', featureId = input$mymap_geojson_mouseover$featureId,
+                              style = list(weight=5,color="black")) # can use a list
+  })
+  observeEvent(input$mymap_geojson_mouseout, {
+        leafletProxy("mymap") %>%
+          styleFeatureGeoJSON(layerId ='geojsonlayer', featureId = input$mymap_geojson_mouseout$featureId,
+                              style = '{"weight": 1, "color": "#555555"}') # or string
+  })
+  observeEvent(input$addBermuda, {
+    # geoJSON layer must already be added
+    # simplified example, not added to addedData
+    leafletProxy("mymap") %>%
+      addFeatureGeoJSON(as.character(minify(bermudaTriangle)), layerId ='geojsonlayer') # can use a GeoJSON string (as.character)
+  })
+  observeEvent(input$removeBermuda, {
+    leafletProxy("mymap") %>%
+      removeFeatureGeoJSON(layerId ='geojsonlayer', featureId = "Bermuda Triangle")
+  })
+}
+shinyApp(ui, server)
+
```

Comment:
Yes it's slow. We really need a way to send over setStyle or resetStyle functions to the client, although I'm not sure yet how to do that and allow customization. Note: I don't believe resetStyle will work for when you effectively want to change back to the prior style instead of the originally loaded style (when you have dynamic data). Internally, on `leaflet.js` we have hard-coded:

```javascript
    function highlightFeature(e) {
    	var layer = e.target;
			layer.setStyle({weight: 3,color: "black"});
	if (!L.Browser.ie && !L.Browser.opera) {
		layer.bringToFront();};
    };
    function resetHighlight(e) {
      var layer = e.target;
      var originalColor = e.target.feature.properties.style.color;
      var originalWeight = e.target.feature.properties.style.weight;
      layer.setStyle({weight: originalWeight,color: originalColor});
    };
```

and inside the `onEachFeature` of `var gjlayer` of `methods.addGeoJSON` we include:

```javascript
    layer.on("mouseover", highlightFeature , this);
    layer.on("mouseout", resetHighlight, this);
```

The capability of assigning styles to events could probably be added to `addGeoJSON`. Then, we probably need some sort of function `styleGeoJSON` that lets you restyle an entire layer (many users won't actually have dynamic data), and a function `styleEventGeoJSON` that lets you restyle the event styles.

ADDED: It looks like the changes you suggested sped it up considerably, thanks! Within the "styled events" we also need a way to send features to the top using `bringToFront` as in this example (if you look at the borders, some are thinner because they are below other features).

Despite being faster, it does still get suck with some highlighted features that you have mouseout'd on (it does not do this on the hard-coded JS version).

UPDATE: moved this comment to here